### ISSUE TITLE
Packages in binary cache do not have full_hash even after re-upload.

### ISSUE THREAD

**Issue by haampie** [LLNL/spack#19665]
Ping @scottwittenburg. After #19209 my CI script constantly uploads all binaries, even if they were downloaded from the same binary cache. My CI scripts ends with:

```
spack buildcache check --rebuild-on-error --mirror-url s3://spack -s /[some_hash]

if [ $? -eq 1 ]; then
    spack buildcache create -af --mirror-url s3://spack --only=package /[some_hash]
fi
```

and currently it always uploads everything, saying 

```
==> Rebuilding [spec]/[some_hash], reason: full_hash was missing from remote spec.yaml
```

I've reran the CI job a couple times, and it will always download binaries from cache, build nothing from sources, and subsequently push everything redundantly to the binary cache again.

Indeed when downloading the spec.yaml files I don't see any `full_hash` entries.

When the user does not require a full hash match, I would expect `buildcache check` to ignore missing `full_hash`es on the remote.

---

Edit, I'll just try to add `--require-full-hash-match` to my CI-script and the problem should be gone.

**Comment by haampie** [comment]
Thinking about this a little bit more, should `spack buildcache check` have a `--require-full-hash-match` too?

**Comment by scottwittenburg** [comment]
I think `spack buildcache check` needs some tests, as it's not a code path I'm using anymore, and as a result it may not have gotten enough scrutiny during development, testing, and review of #19209.  I'll try to reproduce your issue and come up with a test that will keep it working.

**Comment by haampie** [comment]
The tl;dr version is that spack install defaults to not using full_hash, whereas buildcache check defaults to it (or at least returns the package needs to be rebuilt when full_hash is missing)

Better would be to add the same behavior and same flag to buildcache check?

**Comment by haampie** [comment]
@tgamblin: the issue here persists and seems to have gotten worse. Maybe related to #19672? 

For instance, this CI job https://gitlab.com/cscs-ci/electronic-structure/build-spack-images/-/jobs/872101945 was restarted (e.g. https://gitlab.com/cscs-ci/electronic-structure/build-spack-images/-/jobs/872125388), and each time it complains about a full hash mismatch (and uploads 67 packages). There's no other changes to the binary cache in the meantime.

Either the remote full hash remains identical, or the local full hash is simply incorrect. Do you have any idea what's going on?

Worth noting: I've nuked my binary cache right before this.

**Comment by scottwittenburg** [comment]
@haampie It seems we may have finally gotten to the bottom of this.  Let me try to summarize to check if I've got it right.

One issue was you needed to use `--require-full-hash-match` on all your installs to make sure things were rebuilt from source to begin with, before subsequently using `spack buildcache check -s /hash` to see if they should be pushed to the mirror.

But even after that, problems persisted, and it seems the problem was recently compounded by using the new concretizer, which re-concretized already concrete specs, and that issue should be fixed with #20196.

Is that a reasonable summary?

**Comment by haampie** [comment]
Yes, it is, and I've just verified `buildcache check` does the right thing on `develop` now! Thanks for your patience :p
